pkg/logql/log/pipeline.go

```diff
@@ -226,7 +233,12 @@ func (p *streamPipeline) Process(ts int64, line []byte, structuredMetadata ...la
 	p.builder.Reset()
 
 	for i, lb := range structuredMetadata {
-		structuredMetadata[i].Name = prometheus.NormalizeLabel(lb.Name)
+		if cached, ok := p.cache.Load(lb.Name); ok {
+			structuredMetadata[i].Name = cached.(string)
+		} else {
+			structuredMetadata[i].Name = prometheus.NormalizeLabel(lb.Name)
+			p.cache.Store(lb.Name, lb.Name)
+		}
```

Comment:
Hm, I'm not sure about this one. I'm worried this cache will grow pretty large over the lifetime of a query. 

Have you considered a new argument to `p.builder.Add` which specifies to normalize any name it adds? Or a new method like `p.builder.AddUnnormalized`? 

(By the way, the code here is incorrect; it's caching the normalized label name but you're looking up the unnormalized label name from the cache)

Reply:
I agree, I was worried about that. I think its a nice improvement, but we'd need more testing since it also affects the existing query path. I'll remove it from this PR.